x-pack/filebeat/input/httpjson/input.go

```diff
@@ -81,532 +49,150 @@ func newRetryLogger() *retryLogger {
-				if err != nil {
-					return errors.Wrapf(err, "failed to retrieve the next URL for pagination")
-				}
-				if ri.URL == url || url == "" {
-					in.log.Info("Pagination finished.")
-					break
-				}
-				ri.URL = url
-				if err = in.applyRateLimit(ctx, header, in.config.RateLimit); err != nil {
-					return err
-				}
-				in.log.Info("Continuing with pagination to URL: ", ri.URL)
-				continue
-			} else {
-				// Pagination control using HTTP Body fields
-				ri, err = createRequestInfoFromBody(in.config.Pagination, common.MapStr(response), common.MapStr(mm), ri)
-				if err != nil {
-					return err
-				}
-				if ri == nil {
-					break
-				}
-				if err = in.applyRateLimit(ctx, header, in.config.RateLimit); err != nil {
-					return err
-				}
-				in.log.Info("Continuing with pagination to URL: ", ri.URL)
-				continue
-			}
+		switch url.Scheme {
+		case "https":
+			return "443"
 		}
-		break
-	}
+		return "80"
+	}()
 
-	if mm != nil && in.config.DateCursor.IsEnabled() {
-		in.advanceCursor(common.MapStr(mm))
+	_, err = net.DialTimeout("tcp", net.JoinHostPort(url.Hostname(), port), time.Second)
+	if err != nil {
+		return fmt.Errorf("url %q is unreachable", in.config.URL)
 	}
 
 	return nil
 }
 
-func (in *HttpjsonInput) getURL() string {
-	if !in.config.DateCursor.IsEnabled() {
-		return in.config.URL
-	}
+// Run starts the input worker then returns. Only the first invocation
+// will ever start the worker.
+func (in *httpJSONInput) Run(ctx v2.Context, publisher stateless.Publisher) error {
+	log := ctx.Logger.With("url", in.config.URL)
 
-	var dateStr string
-	if in.nextCursorValue == "" {
-		t := timeNow().UTC().Add(-in.config.DateCursor.InitialInterval)
-		dateStr = t.Format(in.config.DateCursor.GetDateFormat())
-	} else {
-		dateStr = in.nextCursorValue
-	}
+	stdCtx := ctxtool.FromCanceller(ctx.Cancelation)
 
-	url, err := url.Parse(in.config.URL)
+	httpClient, err := in.newHTTPClient(stdCtx)
 	if err != nil {
-		return in.config.URL
+		return err
 	}
 
-	q := url.Query()
+	dateCursor := newDateCursorFromConfig(in.config, log)
 
-	var value string
-	if in.config.DateCursor.ValueTemplate == nil {
-		value = dateStr
-	} else {
-		buf := new(bytes.Buffer)
-		if err := in.config.DateCursor.ValueTemplate.Execute(buf, dateStr); err != nil {
-			return in.config.URL
-		}
-		value = buf.String()
-	}
+	rateLimiter := newRateLimiterFromConfig(in.config, log)
 
-	q.Set(in.config.DateCursor.URLField, value)
+	pagination := newPaginationFromConfig(in.config)
 
-	url.RawQuery = q.Encode()
-
-	return url.String()
-}
+	requester := newRequester(
+		in.config,
+		rateLimiter,
+		dateCursor,
+		pagination,
+		httpClient,
+		log,
+	)
 
-func (in *HttpjsonInput) advanceCursor(m common.MapStr) {
-	if in.config.DateCursor.Field == "" {
-		in.nextCursorValue = time.Now().UTC().Format(in.config.DateCursor.GetDateFormat())
-		return
+	if err := requester.processHTTPRequest(stdCtx, publisher); err != nil {
+		return err
```

Comment:
returning here will shut down the input. Is the error that critical that the input is eventually broken or can we recover (e.g. just log and try again after some backoff)?